### PR TITLE
Elide 3.0 add requestscope

### PR DESCRIPTION
**Pull request by tarrantzhang**
#294 Added Requestscope as param for datastoreTranscation functions.

### CODE REVIEW

**Comment by yahoocla** [comment]
CLA is valid!

File: elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java [hunk 1]
```diff
@@ -207,18 +188,77 @@ default void preCommit() {
...
-            Pagination pagination
-    ) {
-        return PersistentResource.getValue(entity, relationName, dictionary);
-    }
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope) {
+        EntityDictionary dictionary = scope.getDictionary();
+        Object val = PersistentResource.getValue(entity, relationName, dictionary);
+        if (val instanceof Collection) {
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: I think we can remove this entire if statement. and just return val above.

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 2]
```diff
@@ -296,8 +299,11 @@ public boolean matchesId(String checkId) {
         }
 
         Iterable<T> list;
-        FilterScope filterScope = new FilterScope(requestScope, loadClass);
-        list = tx.loadObjects(loadClass, filterScope);
+        Optional<FilterExpression> permissionFilter;
+        permissionFilter = requestScope.getPermissionExecutor().getReadPermissionFilter(loadClass);
+
+        list = (Iterable<T>) tx.loadObjects(loadClass, permissionFilter,
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: There are two sources for filter expressions - permissions and the request filter. This is passing one (the permission filter) but not the other (the request filter).  This should be fixed.

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 3]
```diff
@@ -340,8 +346,11 @@ public boolean matchesId(String checkId) {
         }
 
         Iterable<T> list;
-        FilterScope filterScope = new FilterScope(requestScope, loadClass);
-        list = tx.loadObjectsWithSortingAndPagination(loadClass, filterScope);
+        Optional<FilterExpression> permissionFilter = requestScope.getPermissionExecutor()
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Same comment.  We need to also grab the request filters.

File: elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java [hunk 4]
```diff
@@ -33,30 +33,29 @@ default User accessUser(Object opaqueUser) {
...
-     * Save entity to database table.  Save is called after commit checks have evaluated but before the final
-     * transaction commit.
-     *
-     * @param entity record to save
+     * Save the updated object.
+     * @param entity - the object to save.
+     * @param scope - contains request level metadata.
      */
-    void save(Object entity);
+    void save(Object entity, RequestScope scope);
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: This should be the interface RequestScope (from the security package)

File: elide-core/src/main/java/com/yahoo/elide/core/RequestScope.java [hunk 5]
```diff
@@ -357,6 +357,36 @@ else if (permissionFilter.isPresent()) {
     }
 
     /**
+     * Get the global/cross-type filter expression.
+     * @param loadClass
+     * @param permissionFilter filterExpression from FilterExpressionCheck
+     * @return The global filter expression evaluated at the first load
+     */
+    public Optional<FilterExpression> getLoadFilterExpression(Class<?> loadClass,
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/RequestScope.java]: Can we leave this as it were before?

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 6]
```diff
@@ -129,137 +131,164 @@ public void createObject(Object entity, RequestScope scope) {
...
-            }
-            @SuppressWarnings("unchecked")
-            T record = (T) session.load(loadClass, id);
-            Hibernate.initialize(record);
-            return record;
-        } catch (ObjectNotFoundException e) {
-            return null;
-        }
-    }
+//    @Deprecated
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Remove dead code.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 7]
```diff
@@ -129,137 +131,164 @@ public void createObject(Object entity, RequestScope scope) {
...
+//
+//        if (filterExpression.isPresent()) {
+//            CriterionFilterOperation filterOpn = new CriterionFilterOperation(criteria);
+//            criteria = filterOpn.apply(filterExpression.get());
+//        }
+//
+//        final Pagination pagination = filterScope.getRequestScope().getPagination();
+//
+//        // if we have sorting and sorting isn't empty, then we should pull dictionary to validate the sorting rules
+//        Set<Order> validatedSortingRules = null;
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Same

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 8]
```diff
@@ -129,137 +131,164 @@ public void createObject(Object entity, RequestScope scope) {
...
-    public <T> Iterable<T> loadObjectsWithSortingAndPagination(Class<T> entityClass,
-                                                               FilterScope filterScope) {
-        Criterion securityCriterion = filterScope.getCriterion(NOT, AND, OR);
+    public Iterable<Object> loadObjects(
+            Class<?> entityClass,
+            Optional<FilterExpression> filterExpression,
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope) {
+        ParseTree permissions = scope.getDictionary().getPermissionsForClass(entityClass, ReadPermission.class);
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: This shoudl be done by Elide core - not the data store.

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexTransaction.java [hunk 9]
```diff
@@ -193,10 +178,10 @@ protected DataStoreTransaction getTransaction(Class<?> cls) {
                 relationClass, dictionary, filters, sorting, pagination);
     }
 
-    @Override
-    public <T> Iterable<T> loadObjectsWithSortingAndPagination(Class<T> entityClass, FilterScope filterScope) {
-        return getTransaction(entityClass).loadObjectsWithSortingAndPagination(entityClass, filterScope);
-    }
+//    @Override
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexTransaction.java]: Remove dead code.

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexReadTransaction.java [hunk 10]
```diff
@@ -24,12 +24,12 @@ protected DataStoreTransaction beginTransaction(DataStore dataStore) {
...
 
     @Override
-    public void save(Object object) {
+    public void save(Object entity, com.yahoo.elide.core.RequestScope scope) {
         throw new UnsupportedOperationException();
     }
 
     @Override
-    public void delete(Object object) {
+    public void delete(Object entity, com.yahoo.elide.core.RequestScope scope) {
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexReadTransaction.java]: What about getRelation and getAttribute?

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java [hunk 11]
```diff
@@ -86,19 +86,19 @@ public void commit() {
...
                     if (cloned == NEWLY_CREATED_OBJECT) {
-                        transaction.delete(dirtyObject);
+                        transaction.delete(dirtyObject, requestScope);
                     } else {
-                        transaction.save(cloned);
+                        transaction.save(cloned, requestScope);
                     }
                 }
-                transaction.commit();
+                transaction.commit(null);
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java]: Why null?

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java [hunk 12]
```diff
@@ -171,38 +171,23 @@ private Object cloneObject(Object object) {
...
-        if (relation instanceof Iterable) {
-            return hold(transaction, (Iterable) relation);
-        }
-
-        return hold(transaction, relation);
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope) {
+        DataStoreTransaction transaction = getTransaction(entityClass);
+        return hold(transaction, transaction.loadObjects(entityClass, filterExpression, sorting, pagination, scope));
```
Comment (aklish) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java]: We should add get/setRelation and get/setAttribute

File: elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java (hunk 4 above)
Comment (tarrantzhang) [inline elide-core/src/main/java/com/yahoo/elide/core/DataStoreTransaction.java]: I changed the import, so it's security.requestscope now.

File: elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java (hunk 12 above)
Comment (tarrantzhang) [inline elide-datastore/elide-datastore-multiplex/src/main/java/com/yahoo/elide/datastores/multiplex/MultiplexWriteTransaction.java]: I will add implementation of this part to a new PR.

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 13]
```diff
@@ -296,8 +299,10 @@ public boolean matchesId(String checkId) {
         }
 
         Iterable<T> list;
-        FilterScope filterScope = new FilterScope(requestScope, loadClass);
-        list = tx.loadObjects(loadClass, filterScope);
+        Optional<FilterExpression> filterExpression = requestScope.getLoadFilterExpression(loadClass);
+
+        list = (Iterable<T>) tx.loadObjects(loadClass, filterExpression,
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Don't we need to get the FilterExpression for the permissions and join them with the request filter expressions?

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 14]
```diff
@@ -340,8 +345,10 @@ public boolean matchesId(String checkId) {
         }
 
         Iterable<T> list;
-        FilterScope filterScope = new FilterScope(requestScope, loadClass);
-        list = tx.loadObjectsWithSortingAndPagination(loadClass, filterScope);
+        Optional<FilterExpression> filterExpression = requestScope.getLoadFilterExpression(loadClass);
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Same comment.  We need to merge in the permission filters.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 15]
```diff
@@ -129,125 +131,81 @@ public void createObject(Object entity, RequestScope scope) {
...
+            Class<?> entityClass,
+            Optional<FilterExpression> filterExpression,
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope) {
+        try {
+            com.yahoo.elide.core.RequestScope requestScope = (com.yahoo.elide.core.RequestScope) scope;
+            ParseTree permissions = requestScope.getDictionary()
+                    .getPermissionsForClass(entityClass, ReadPermission.class);
+            Criterion securityCriterion = requestScope.getPermissionExecutor().getCriterion(permissions, NOT, AND, OR);
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: I'm thinking we can remove this and remove the concept of Criteria checks.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 16]
```diff
@@ -393,37 +351,39 @@ public T getObject() {
...
+                            .withPossibleSorting(sorting)
+                            .withPossiblePagination(pagination)
+                            .build();
+                    if (possibleQuery.isPresent()) {
+                        return possibleQuery.get().list();
+                    }
                 }
             }
+            return val;
+        } catch (ClassCastException e) {
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Keep this catch block closer to the cast.  The only statement in the try should be the cast.

File: elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java [hunk 17]
```diff
@@ -486,7 +446,7 @@ public User accessUser(Object opaqueUser) {
     }
 
     @Override
-    public void setRequestScope(RequestScope requestScope) {
+    public void setRequestScope(com.yahoo.elide.core.RequestScope requestScope) {
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Why is this core and not security?

Comment (aklish) [inline elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java]: Another question - why does this method exist given that every method takes request scope?  Can we just remove this method?

File: elide-datastore/elide-datastore-hibernate5/src/main/java/com/yahoo/elide/datastores/hibernate5/HibernateTransaction.java [hunk 18]
```diff
@@ -121,115 +121,89 @@ public void createObject(Object entity, RequestScope scope) {
...
+            Class<?> entityClass,
+            Optional<FilterExpression> filterExpression,
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope) {
+        try {
+            com.yahoo.elide.core.RequestScope requestScope = (com.yahoo.elide.core.RequestScope) scope;
+            ParseTree permissions = requestScope.getDictionary()
+                    .getPermissionsForClass(entityClass, ReadPermission.class);
+            Criterion securityCriterion = requestScope.getPermissionExecutor()
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate5/src/main/java/com/yahoo/elide/datastores/hibernate5/HibernateTransaction.java]: I think we likely can remove all of this logic.

File: elide-datastore/elide-datastore-hibernate5/src/main/java/com/yahoo/elide/datastores/hibernate5/HibernateTransaction.java [hunk 19]
```diff
@@ -121,115 +121,89 @@ public void createObject(Object entity, RequestScope scope) {
...
 
-        if (filterExpression.isPresent()) {
-            CriterionFilterOperation filterOpn = new CriterionFilterOperation(criteria);
-            criteria = filterOpn.apply(filterExpression.get());
+            return loadObjects(
+                    entityClass,
+                    criteria,
+                    Optional.ofNullable(validatedSortingRules),
+                    pagination);
+        } catch (ClassCastException e) {
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate5/src/main/java/com/yahoo/elide/datastores/hibernate5/HibernateTransaction.java]: Move this closer to the cast.

File: elide-datastore/elide-datastore-hibernate5/src/main/java/com/yahoo/elide/datastores/hibernate5/HibernateTransaction.java [hunk 20]
```diff
@@ -249,36 +223,40 @@ public void createObject(Object entity, RequestScope scope) {
...
+                            .withPossibleSorting(sorting)
+                            .withPossiblePagination(pagination)
+                            .build();
+                    if (possibleQuery.isPresent()) {
+                        return possibleQuery.get().list();
+                    }
                 }
             }
+            return val;
+        } catch (ClassCastException e) {
```
Comment (aklish) [inline elide-datastore/elide-datastore-hibernate5/src/main/java/com/yahoo/elide/datastores/hibernate5/HibernateTransaction.java]: Move closer to cast.

**Comment by tarrantzhang** [comment]
I also removed filterScope and functions and class related to criterionCheck.

**Comment by aklish** [comment]
👍

**Pull request merged**